subsys/net/ip/net_shell.c

```diff
@@ -3534,159 +3753,249 @@ static void iface_vlan_cb(struct net_if *iface, void *user_data)
+		  cmd_net_http_monitor),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_iface)
+{
+	SHELL_CMD(up, NULL, "Take network interface up.", cmd_net_iface_up),
+	SHELL_CMD(down, NULL, "Take network interface down.",
+		  cmd_net_iface_down),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_nbr)
+{
+	SHELL_CMD(rm, NULL, "Remove neighbor from cache.",
+		  cmd_net_nbr_rm),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_stats)
+{
+	SHELL_CMD(all, NULL,
+		  "Show network statistics for all network interfaces.",
+		  cmd_net_stats_all),
+	SHELL_CMD(iface, NULL,
+		  "Show network statistics for one network interface.",
+		  cmd_net_stats_iface),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_tcp)
+{
+	SHELL_CMD(connect, NULL, "Connect to TCP peer.", cmd_net_tcp_connect),
+	SHELL_CMD(send, NULL, "Send data to peer using TCP.",
+		  cmd_net_tcp_send),
+	SHELL_CMD(close, NULL, "Close TCP connection.", cmd_net_tcp_close),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_vlan)
+{
+	SHELL_CMD(add, NULL, "Add VLAN tag to the network interface.",
+		  cmd_net_vlan_add),
+	SHELL_CMD(del, NULL, "Delete VLAN tag from the network interface.",
+		  cmd_net_vlan_del),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_commands)
+{
+	/* Alphabetically sorted. */
+	SHELL_CMD(allocs, NULL, "Print network memory allocations.",
+		  cmd_net_allocs),
+	SHELL_CMD(app, NULL,
+		  "Print network application API usage information.",
+		  cmd_net_app),
+	SHELL_CMD(arp, &net_cmd_arp, "Print information about IPv4 ARP cache.",
+		  cmd_net_arp),
+	SHELL_CMD(conn, NULL, "Print information about network connections.",
+		  cmd_net_conn),
+	SHELL_CMD(dns, &net_cmd_dns, "Show how DNS is configured.",
+		  cmd_net_dns),
+	SHELL_CMD(gptp, &net_cmd_gptp, "Print information about gPTP support.",
+		  cmd_net_gptp),
+	SHELL_CMD(http, &net_cmd_http,
+		  "Print information about active HTTP connections.",
+		  cmd_net_http),
+	SHELL_CMD(iface, &net_cmd_iface,
+		  "Print information about network interfaces.",
+		  cmd_net_iface),
+	SHELL_CMD(mem, NULL, "Print information about network memory usage.",
+		  cmd_net_mem),
+	SHELL_CMD(nbr, &net_cmd_nbr, "Print neighbor information.",
+		  cmd_net_nbr),
+	SHELL_CMD(ping, NULL, "Ping a network host.", cmd_net_ping),
+	SHELL_CMD(route, NULL, "Show network route.", cmd_net_route),
+	SHELL_CMD(rpl, NULL, "Show RPL mesh routing status.", cmd_net_rpl),
+	SHELL_CMD(stacks, NULL, "Show network stacks information.",
+		  cmd_net_stacks),
+	SHELL_CMD(stats, &net_cmd_stats, "Show network statistics.",
+		  cmd_net_stats),
+	SHELL_CMD(tcp, &net_cmd_tcp, "Connect/send/close TCP connection.",
+		  NULL),
+	SHELL_CMD(vlan, &net_cmd_vlan, "Show VLAN information.", cmd_net_vlan),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CMD_REGISTER(net, &net_commands, "Networking commands", NULL);
+
+SHELL_UART_DEFINE(shell_transport_uart);
```

Comment:
you should leave shell definition to the application. Here only commands should be specified.

Reply:
Hmm, with net-shell the idea is that the app only needs to set CONFIG_NET_SHELL=y and get a shell automatically.

Reply:
note, that in future there might be multiple transports (net too) and even multiple shell instances available so that approach will not fly for long. We might approach that the other way: add UART as default transport to the shell module.

Reply:
@nordic-krch The `SHELL_DEFINE()` macro is a bit problematic as it requires the transport as a parameter. As a user I would just like to set `CONFIG_NET_SHELL=y` and expecting that net-shell is enabled without any configuration. Could the transport parameter left out or set to NULL? The shell subsystem would then just select the default backend in that case. Setting the backend here looks something that would be better done in some other part of the system.